server/src/main/java/io/druid/server/QueryResource.java

```diff
@@ -220,8 +222,17 @@ public void write(OutputStream outputStream) throws IOException, WebApplicationE
                 },
                 contentType
             )
-            .header("X-Druid-Query-Id", queryId)
-            .header("X-Druid-Response-Context", jsonMapper.writeValueAsString(responseContext))
+            .header("X-Druid-Query-Id", queryId);
+
+        //Limit the response-context header, see https://github.com/druid-io/druid/issues/2331
+        String responseCtxString = jsonMapper.writeValueAsString(responseContext);
+        if (responseCtxString.length() > RESPONSE_CTX_HEADER_LEN_LIMIT) {
+          log.warn("Response Context truncated for id [%s] . Full context is [%s].", queryId, responseCtxString);
+          responseCtxString = responseCtxString.substring(0, RESPONSE_CTX_HEADER_LEN_LIMIT);
```

Comment:
IMO there should be another header to tell you whether the response context has been truncated, otherwise callers will have to rely on catching parse errors, which may mask other problems